Tasks/ServiceFabricDeployV1/ServiceFabricSDK/Publish-UpgradedServiceFabricApplication.ps1

```diff
@@ -372,12 +372,80 @@ function Publish-UpgradedServiceFabricApplication
         }
 
         Write-Host (Get-VstsLocString -Key SFSDK_WaitingForUpgrade)
-        $upgradeStatusFetcher = { Get-ServiceFabricApplicationUpgradeAction -ApplicationName $ApplicationName }
+        $unhealthyEvaluationMessageFunction = {
+            param(
+                $UnhealthyEvaluations,
+
+                [string]$Indentation
+            )
+
+            if ($UnhealthyEvaluations -eq $null) {
+                return ""
+            }
+
+            $indentatedErrorString = ""
+            foreach ($UnhealthyEvaluation in $UnhealthyEvaluations) {
+                $indentatedErrorString += $Indentation + $UnhealthyEvaluation.Description
+                # see if indentation needs to be increased. based on the type of evaluation.
+                $indentatedErrorString += & $unhealthyEvaluationMessageFunction $UnhealthyEvaluation, if($UnhealthyEvaluations.kind -eq ($UnhealthyEvaluation.kind + "s")) { return $Indentation } Else { return $Indentation + "`t" }
+            }
+
+            if ($UnhealthyEvaluation.UnhealthyEvent) {
+                $indentatedErrorString += $Indentation + $UnhealthyEvaluation.UnhealthyEvent
+            }
+
+            return $indentatedErrorString
+        }
+
+        $DomainUpgradeStatusFormatterFunction = {
+            param(
+                $UpgradeDomainsStatus
+            )
+
+            if ($UpgradeDomainsStatus -eq $null) {
+                return ""
+            }
+
+            $enumerator = $UpgradeDomainsStatus.GetEnumerator()
+            $upgradeDomainStatusString = ""
+            while ($enumerator.MoveNext() -ne $False) {
```

Comment:
Can you not use "ConvertTo-Json -InputObject $UpgradeDomainsStatus -Compress" directly?